### PR TITLE
[Validator] Add agent_state validator

### PR DESCRIPTION
**Pull request by Skylion007**
## Motivation and Context

As a sanity check, prevent users from setting an agent_state which has inf or NaN in any of the arrays. It is not sensible behavior and is indicative of an error. 

## How Has This Been Tested
* This has been tested with PyTest

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: habitat_sim/agent/agent.py [hunk 1]
```diff
@@ -54,18 +59,20 @@ class SixDOFPose(object):
...
-    angular_velocity: np.ndarray = np.zeros(3)
-    force: np.ndarray = np.zeros(3)
-    torque: np.ndarray = np.zeros(3)
+    velocity: np.ndarray = attr.ib(factory=_triple_zero, validator=all_is_finite)
+    angular_velocity: np.ndarray = attr.ib(
+        factory=_triple_zero, validator=all_is_finite
+    )
+    force: np.ndarray = attr.ib(factory=_triple_zero, validator=all_is_finite)
+    torque: np.ndarray = attr.ib(factory=_triple_zero, validator=all_is_finite)
     sensor_states: Dict[str, SixDOFPose] = attr.Factory(dict)
```
Comment (erikwijmans) [inline habitat_sim/agent/agent.py:76]: Can you add similar checks to `SixDOFPose`?

File: habitat_sim/agent/agent.py [hunk 2]
```diff
@@ -54,18 +59,20 @@ class SixDOFPose(object):
...
-    position: np.ndarray = np.zeros(3)
+    position: np.ndarray = attr.ib(factory=_triple_zero, validator=all_is_finite)
     rotation: Union[np.quaternion, List] = np.quaternion(1, 0, 0, 0)
 
 
 @attr.s(auto_attribs=True, slots=True)
 class AgentState(object):
-    position: np.ndarray = np.zeros(3)
+    position: np.ndarray = attr.ib(factory=_triple_zero, validator=all_is_finite)
     rotation: Union[np.quaternion, List, np.ndarray] = np.quaternion(1, 0, 0, 0)
```
Comment (erikwijmans) [inline habitat_sim/agent/agent.py:69]: Let's also check for the rotation being unit length. `.norm()` for `np. np.quaternion` and `np.linalg.norm` for the others.

Comment (Skylion007) [reply habitat_sim/agent/agent.py:69]: Are they any np.quaternion functions to do this @erikwijmans ? A lot of numpy functions like np.finite don't work on that type.

File: habitat_sim/agent/agent.py (hunk 1 above)
Comment (Skylion007) [reply habitat_sim/agent/agent.py:76]: @erikwijmans I do in the SixDOFPose struct itself.

File: habitat_sim/agent/agent.py (hunk 2 above)
Comment (erikwijmans) [reply habitat_sim/agent/agent.py:69]: `.norm()` will do it.  That will not be 1 when there is inf or nan

File: habitat_sim/agent/agent.py (hunk 1 above)
Comment (erikwijmans) [reply habitat_sim/agent/agent.py:76]: Yeah, that is just another attrs class defined right above this one

Comment (Skylion007) [reply habitat_sim/agent/agent.py:76]: Not sure what you are asking then?

File: habitat_sim/utils/validators.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+def all_is_finite(instance, attribute, value) -> None:
+    if not np.all(np.isfinite(value)):
+        raise ValueError(
+            f"{value} contains non-finite values which are forbidden in the {attribute} of {instance}"
+        )
+
+
+def is_unit_length(instance, attribute, value, tol=1e-5) -> None:
+    if isinstance(value, mn.Quaternion):
```
Comment (erikwijmans) [inline habitat_sim/utils/validators.py:24]: mn.Quaternion has https://doc.magnum.graphics/python/magnum/Quaternion/#is_normalized

**Review by erikwijmans — CHANGES_REQUESTED** [review]
Looking at the regression test on the CI, it doesn't seem like this is causing a large enough one to cause that to error, but this does seem to be slowing things down by 10-20%.  We call `Agent.get_state` on every frame currently, however the state constructed in that should never fail the validators (other things would have errored long before in the renderer).

Looks like attrs validators can be globally disabled via `attr.set_run_validators(False)` (and then re-enabled via `attr.set_run_validators(True)`).  Can you add that to get_state to keep things fast when already know things will be valid?

**Comment by Skylion007** [comment]
Validators are only run on state construction (from __init__) and when
attr.validate is called, so I am not sure that would fix the slowdown. Are
you sure get_state is the culprit? Using pyspy?

**Comment by Skylion007** [comment]
@eundersander Could be useful for some of your profiling tools here.

**Comment by erikwijmans** [comment]
Yeah, I looked at current master vs. this PR.  attrs validators are also run on construction for default values.

**Comment by Skylion007** [comment]
Ah, didn't realized they ran on defaults

**Comment by erikwijmans** [comment]
> Ah, didn't realized they ran on defaults

Neither did I, had to double check :-)

**Comment by Skylion007** [comment]
@erikwijmans This PR did catch a potential bug. 

If initial_state of an agent is None, there is a chance that the sample_random_navigable_point will return an invalid state.

**Comment by Skylion007** [comment]
@erikwijmans Still looks like a 10-20% slowdown...

**Comment by erikwijmans** [comment]
> @erikwijmans Still looks like a 10-20% slowdown...

I am not seeing that locally.  If I compare this and master back to back, all seems about equal.  If I comment out the validator disable, then I see the perf drop.

**Comment by mathfac** [comment]
I find random position sampling for the agent useful and was actually using None starting position/rotation as feature.
For example if user would like to have an episode where agent position is random? Sampling position to fill the episode always require loading the scene first.

Possibly easier way to proceed will be to check sampled start position when it wasn't provided?

**Comment by Skylion007** [comment]
@mathfac @erikwijmans bump? Only thing I don't quite like is doing the deep validation of the dict of sensor poses every time.

**Review by erikwijmans — APPROVED** [review]
This seems fine with me

**Pull request merged**